docs/content/sips/0xx-spin-factors.md

```diff
@@ -0,0 +1,108 @@
+Spin codebase. This not only makes it hard to read and understand these features
+but also raises the bar for unfamiliar developers trying to add new features.
+
+Separately, the introduction of the SpinKube project has made it clear that
+there will be (at least) two major embeddings of the Spin runtime environment:
+the Spin CLI (`spin up`) and SpinKube (`containerd-shim-spin`). In order to
+provide better integration into the Kubernetes ecosystem, some runtime feature
+implementations will diverge between these embeddings, making the loose coupling
+of features even more important.
+
+## Proposal
+
+The basic inversion of control approach used by `HostComponent`s will be
+redesigned and expanded into a new system called Spin Factors, where independent
+feature sets are organized into individual "factors". Some of the goals of this
+new system:
+
+- Expand the set integration points that a factor can use to hook into the
+  lifecycle of an application request. These will be driven by feature
+  requirements and may include e.g.:
+  - Runtime startup
+  - Application initialization
+  - Runtime config parsing
+  - Component pre-instantiation
+  - Component instantiation
+  - Cleanup (post-execution)
+
+- Allow loosely-coupled dependencies between "factors" (features). In order to
+  simplify implementation and reasoning about these dependencies the dependency
+  graph must be acyclic.
+
+## Implementation Plan
+
+The overall implementation plan will be to (in parallel):
+
+- Introduce a new `spin-factors` crate containing the basic framework.
+- Refactor existing Spin features (mostly existing `HostComponent`s) to use
+   this new framework.
+   - Framework features will be developed as needed to support this refactoring.
+- Refactor `spin-core`, `spin-trigger`, etc. to use `spin-factors`.
+   - Depending on subjective evaluation at this point, possibly merge
+   `spin-core` and `spin-factors`.
+
+## Implementation Details
+
+Based on initial prototyping, the following Rust types represent the starting
+point for `spin-factors` (note that some type details are elided for clarity):
+
+```rust
+pub trait Factor {
+    // Builder represents a type that can prepare request state for the factor.
+    type Builder: FactorBuilder<Self>;
+    // Data represents request state for the factor.
+    type Data;
+
+    // Init is the runtime startup lifecycle hook.
+    //
+    // The `InitContext` type here gives the factor the ability to update global
+    // engine configuration, most notably the `Linker`. This takes the place of
+    // `HostComponent::add_to_linker`.
+    fn init(ctx: InitContext) -> Result<()> {
+        Ok(())
+    }
+
+    // Validate app is the application initialization hook.
+    //
+    // This takes the place of `DynamicHostComponent::validate_app`.
+    fn validate_app(&self, app: &App) -> Result<()> {
+        Ok(())
+    }
+}
+
+pub trait FactorBuilder<Factor> {
```

Comment:
`FactorBuilder` makes it sound like this builds a `Factor` but in fact it builds a `Factor::Data` which is meant to represent "request state" - I think some adjustments to naming would be helpful.

Reply:
I'm also curious about the choice of making `FactorBuilder` a generic type instead of an associated type. I would imagine that each implementation of `FactorBuilder` would have 1 and only 1 `Factor` it's associated with, no?

Reply:
I believe it was to support this: https://github.com/lann/spin-factors/blob/cef37d031bbda20e8c0894222be82b4e7d3a85e8/src/lib.rs#L187-L189

```rust
impl<T: Factor> FactorBuilder<T> for DefaultBuilder
where
    T::Data: Default,
```